File: topic1/problem15/Solution.py
class Solution:
    def candy(self, ratings: List[int]) -> int:
        N = len(ratings)
        candies = [1] * N

        for i in range(1, N):
            if ratings[i-1] < ratings[i]:
                candies[i] = candies[i-1] +1
        
        total_candies = candies[-1]

        for i in range(N-2, -1, -1):
            if (ratings[i+1] < ratings[i] and candies[i] <= candies[i+1]):
                candies[i] = candies[i+1] +1

            total_candies += candies[i]
        
        return sum(candies)

File: topic1/problem15/Solution.java
class Solution {
    public int candy(int[] ratings) {
        int N = ratings.length;
        int[] candies = new int[N];

        for (int i = 0; i < N; i++)
            candies[i] = 1;

        for (int i = 1; i < N; i++)
            if (ratings[i] > ratings[i-1])
                candies[i] = candies[i-1] +1;
        
        int totalCandies = candies[N-1];
        for (int i = N-2; i >= 0; i--) {
            if ((ratings[i] > ratings[i+1]) && (candies[i] <= candies[i+1]))
                candies[i] = candies[i+1] +1;
            totalCandies += candies[i];
        }

        return totalCandies;
    }
}
